src/core/compact_object.cc

```diff
@@ -1271,6 +1277,33 @@ void CompactObj::Reset() {
   mask_ = 0;
 }
 
+uint8_t CompactObj::GetFirstByte() const {
+  if (IsInline()) {
+    return u_.inline_str[0];
+  }
+
+  if (taglen_ == ROBJ_TAG) {
+    CHECK_EQ(OBJ_STRING, u_.r_obj.type());
+    DCHECK_EQ(OBJ_ENCODING_RAW, u_.r_obj.encoding());
+    return *(uint8_t*)u_.r_obj.inner_obj();
+  }
+
+  if (taglen_ == SMALL_TAG) {
+    return u_.small_str.first_byte();
+  }
+
+  if (taglen_ == EXTERNAL_TAG) {
+    if (u_.ext_ptr.is_cool) {
+      const CompactObj& cooled_obj = u_.ext_ptr.cool_record->value;
+      return cooled_obj.GetFirstByte();
+    }
```

Comment:
At all time the cooled object is equal to the tiered one, so it could be just ext_ptr.first_byte

Reply:
we currently do not set `u_.ext_ptr.first_byte` for cooled objects, only for fully external.
I was not sure if we can get out of sync, so preferred to read from the "original".